holoviews/plotting/bokeh/element.py

```diff
@@ -168,18 +166,56 @@ def __init__(self, element, plot=None, **params):
         self.current_ranges = None
         super(ElementPlot, self).__init__(element, **params)
         self.handles = {} if plot is None else self.handles['plot']
-        element_ids = self.hmap.traverse(lambda x: id(x), [Element])
-        self.static = len(set(element_ids)) == 1 and len(self.keys) == len(self.hmap)
+        self.static = len(self.hmap) == 1 and len(self.keys) == len(self.hmap)
+        self.callbacks = self._init_callbacks()
 
 
-    def _init_tools(self, element):
+    def _init_callbacks(self):
+        """
+        Initializes any callbacks for streams which have defined
+        the plotted object as a source.
+        """
+        if not self.static or isinstance(self.hmap, DynamicMap):
+            source = self.hmap
+        else:
+            source = self.hmap.last
+        streams = Stream.registry.get(source, [])
+        registry = Stream._callbacks['bokeh']
+        callbacks = {(registry[type(stream)], stream) for stream in streams
+                     if type(stream) in registry and streams}
+        cbs = []
+        for cb, group in groupby(sorted(callbacks), lambda x: x[0]):
+            cb_streams = [s for _, s in group]
+            cbs.append(cb(self, streams, source))
+        return cbs
+
+
+    def _init_tools(self, element, callbacks=[]):
         """
         Processes the list of tools to be supplied to the plot.
         """
-        tools = self.default_tools + self.tools
+        if self.batched:
+            dims = self.hmap.last.kdims
+        else:
+            dims = list(self.overlay_dims.keys())
+        dims += element.dimensions()
+        tooltips = [(d.pprint_label, '@'+util.dimension_sanitizer(d.name))
+                    for d in dims]
+
+        callbacks = callbacks+self.callbacks
+        cb_tools = []
+        for cb in callbacks:
+            for handle in cb.handles:
+                if handle and handle in known_tools:
```

Comment:
If I am following this right, this could be simplified if there was a single (custom) tool for capturing events? I believe this code is about finding the right tool that the callback needs...